### ISSUE TITLE
North Division Street through Spokane isn't labeled as such

### ISSUE THREAD

**Issue by Carnildo** [ZeLonewolf/openstreetmap-americana#321]
https://zelonewolf.github.io/openstreetmap-americana/#15.09/47.70128/-117.41106

North Division Street, the main north-south road through Spokane, is not labeled as such anywhere at any zoom level.  It has plenty of "US-2" and "US-395" highway shields, but no name.

(Interestingly, I see occasional flashes of what might be a name during transitions as I pan or zoom the map, but nothing permanent.)

**Comment by ZeLonewolf** [comment]
I suspect that there is some kind of collision happening here, and I also notice on a few nearby streets that name labels never show up, no matter how close in you zoom.  There may also be tile boundaries in play.

Additionally, until onthegomap/planetiler#14 is resolved, the tile server will continue to have fully-spelled out road names, for example "North Division Street" instead of "N Division St".  The shorter name labels should help reduce the number of collisions with other objects.

**Comment by 1ec5** [comment]
> I suspect that there is some kind of collision happening here, and I also notice on a few nearby streets that name labels never show up, no matter how close in you zoom.

Correct, the issue is that both the `road_label` and `highway-shield` layers are using the same default `symbol-spacing` and `text-offset`. Both layers are trying to center their symbols on the same coordinates; the `highway-shield` layer always wins out because it’s on top. Here’s what the street looks like with `showCollisionBoxes` enabled:

<img src="https://user-images.githubusercontent.com/1231218/219273136-179aa23a-f766-4aa6-9a23-5f5e9cea0e68.png" width="400" alt="North Division Street">

This issue affects many but not all roads. It depends on the visual length of the road name label and that of the concurrent shields when laid out along the road.

**Comment by ZeLonewolf** [comment]
Slack discussion: https://osmus.slack.com/archives/C01V02K52UX/p1676521366781129

**Comment by ZeLonewolf** [comment]
This discussion makes me wonder if we actually want to introduce randomness in the standoff distances on labels rather than fully deterministic ones.

**Comment by 1ec5** [comment]
There’s no provision for explicit randomness in expressions – https://github.com/mapbox/mapbox-gl-js/issues/5853#issuecomment-351448862 – let alone any alternation between two symbols of a single geometry in a single layer. If we just want to ensure that a shield and a label are always present when available, then offsetting one of the layers should be enough to ensure that visibility. To avoid a situation where otherwise centered labels would appear uncentered, we could add a redundant label layer with the offset.

**Comment by ZeLonewolf** [comment]
> There’s no provision for explicit randomness in expressions

This would be a near-trivial thing to add to maplibre-gl-js, and let's not let the lack of expression support limit our thinking. It's an acceptable answer if random offsets isn't the right solution, but I'd prefer to keep the solution space broad as a general rule.

**Comment by 1ec5** [comment]
The lack of a random expression operator isn’t an artificial limitation. The obvious implementation using `random()` most likely wouldn’t behave as you’d expect, especially when caching is involved.
